### DISCUSSION TITLE
New MRXS problem: openslide2vips: opening slide: Cannot read slide position info: Expected 1 value

### DISCUSSION THREAD

**Discussion by tvngithub** [libvips/libvips#3902] [Help wanted]
I am using vips tools to convert MRXS to Zoomify using dzsave on the Ubuntu platform.
The known error in OpenSlide in dealing with MRXS format change by 3DHistech
"Cannot read slide position info: Expected 1 value"
could be worked around by using 3DHistech own converter tool to convert MRXS to MRXS.
However the latest MRXS file we've received from Budapest produced by their newest scanner cannot be worked around with their own converter anymore. Their engineers have introduced new changes that their converter can not deal with. The solution is to use OpenSlide 4.0 to read the file and that works.
However the vips tools release  on Ubuntu platform is still built with the previous version of OpenSlide and thus I have the error above when trying to use vips dzsave. Since 3DHistech converter can no longer be a work around, I am looking for a depository of vips tools built with OpenSlide 4.0 for the Ubuntu platform, and do not want to build my own if I don't have to. 
BTW, 3DHistech engineering is very uncooperative in helping their customers because they have a close format and behave as monopoly. Unfortunately we have been developing software based on their scanner and have to put up with surprised changes.
Thank you very much for any hint.
Best Regards.
-tom

**Comment by jcupitt** [comment]
Hi @tvngithub,

If you're OK with a GUI tool, vipsdisp includes the latest openslide and libdicom:

https://github.com/jcupitt/vipsdisp

You can install the binary with flatpak. Load your MRXS, then save-as `xxx.dz` and set the layout to zoomify.

If you need a command-line converter, you'll need to build your own, I think. It's easy to deploy via docker / dockerhub and have a single binary you use everywhere.

**Comment by tvngithub** [comment]
Thank you very much, John.
I will try the GUI tool first then build the command line tool to use it in
the script.
Best regards.
-tom

**Comment by tvngithub** [comment]
It seems that the GUI version AND my own build of release version 8.13.3 only give exactly ONE TileGroup0 and not the entire image as I used to have with vips dzsave. The newest release 8.15.2 does not have dzsave anymore. This is a huge setback for us in using vips to convert MRXS to Zoomify. It used to work beautifully to convert MRXS to Zoomify.
Now that  3DHistech is dealing us with a change in their format and vips doesn't have dzsave anymore, the alternative is to write our own OpenSlide based conversion. That would take time AND the code may not be as efficient as vips as John and developers have spent lot of time optimizing it.
I am fairly stuck at this point until we write our own converter.
Thank you very much John for your great contribution.
Best Regards.
-tom

**Reply by jcupitt** [reply]
No no, dzsave is still there, in fact it should be noticeably quicker. We switched the zip library from libgsf to libarchive, perhaps you didn't add that package?

**Reply by jcupitt** [reply]
Did vipsdisp fail to display the image? It's just using openslide4, so any openslide4 image ought to work.

**Comment by tvngithub** [comment]
When I run the 8.15.2  tools I got:
`vips --vips-concurrency=3 dzsave F50_20x_calibrated.mrxs[autocrop] ...../F50_20x_calibrated --layout zoomify 
vips: unknown action "dzsave"
` 
When I run the 8.13.3 built I got:
`$vips dzsave F50_20x_calibrated.mrxs[autocrop] F50_20x_calibrated --layout zoomify`
but then there is only one TileGroup0
`$ls F50_20x_calibrated/
ImageProperties.xml  TileGroup0  vips-properties.xml`
That is the same one TileGroup output I received from the GUI tool with this setting
![WEnrBNp9mC0ulgx7](https://github.com/libvips/libvips/assets/25333111/a8e3268b-159a-4829-87ed-e8442d9f5090)

According to this message, dzsave was removed
https://github.com/kleisauke/net-vips/issues/203

So I have problems of generating Zoomify from MRXS either using the GUI tool or the release 8.13.3 that I built.

Thank you.

**Reply by jcupitt** [reply]
We were forced to remove dzsave from our binaries briefly, but we've switched to libarchive and it's back now. It should work fine. Where did you get your libvips binary? If you built it yourself, did you install the libarchive-dev package before configuring libvips?

If the vipsdisp binary can't display your image, it might not be supported by openslide4. Do you have a sample failing MRXS?

**Comment by tvngithub** [comment]
I see and I am happy to see that dzsave is still there.
I built my own vips tools using the instruction from the vips github page. All works well. But I am not sure if I install libarchive-dev package and will check.
The GUI tools displays my MRXS file well but the issue is that it only generates one TIleGroup0 instead of all other as before. That is the same behavior of the 8.13.3 tools I built.
Let me zip up the MRXS file that I a working with and add the file 
The tar.gz file is 2.2GB. Let me see if I can provide a link

**Comment by jcupitt** [comment]
Nice image. It seems to work for me. Using the GUI, I tried saving as zoomify and it made this:

```
$ ls x/F50_20x_calibrated/
ImageProperties.xml  TileGroup1395  TileGroup1793  TileGroup2190  TileGroup601
TileGroup0           TileGroup1396  TileGroup1794  TileGroup2191  TileGroup602
TileGroup1           TileGroup1397  TileGroup1795
...
```

I used these settings:

![image](https://github.com/libvips/libvips/assets/580843/d7ff79b8-3f5c-40f8-81af-031fb8e6bf0f)

It works from the CLI as well with `dzsave`.

**Comment by tvngithub** [comment]
Thank you very much for verifying that vips works with the new 3DHistech format. I am happy that future release on Ubuntu will continue to work for me. However, I am still puzzling about the result of the conversion that I am running. Both GUI and CLI gives exactly the same one TileGroup0. The ImageProperty.xml gives clue why it is

`<IMAGE_PROPERTIES WIDTH="870" HEIGHT="1942" NUMTILES="43" NUMIMAGES="1" VERSION="1.8" TILESIZE="256" />`

For some reason, the converter only see an image of 870x1942 and not the entire MRXS slide. Do you have any clue for me to investigate? I am using the exact settings for the GUI and the CLI is as I posted earlier.
Thanks again for your help.
-tom

**Reply by jcupitt** [reply]
I think your libvips can't have been built with openslide support. It's loading the MRXS file as a JPEG and just seeing a very low-res overview.

When you run `meson setup`, you need to look through the summary table at the end and check it found everything.

**Reply by tvngithub** [reply]
You are exactly right. The meson build does list OpenSlide as not found:

Run-time dependency openslide found: NO (tried pkgconfig and cmake)

So I've been trying to update OpenSlide release to the latest build for Ubuntu 4.0.2  from the source depository. But meson still does not recognize OpenSlide on the system, although all OpenSlide tools work as expected. 
What does VIPS expect from which component of OpenSlide to build with its support? Maybe I should ask this question in the OpenSlide discussion.
Thank you.

**Reply by jcupitt** [reply]
libvips uses `pkg-config` to find openslide, so make sure that's working. 

Test it with:

```
$ pkg-config openslide --modversion
3.4.1
```

It should print the version number of the package.

`pkg-config` looks for a file called `openslide.pc` in the standard locations, and along `PKG_CONFIG_PATH`. Check the prefix you installed openslide to, and make sure that's part of your environment.

**Reply by tvngithub** [reply]
I used pkg-config and it returns OpenSlide version 4.0.0 which is the latest build available for Ubuntu

`$ pkg-config openslide --modversion
4.0.0
`
meson build still doesn't find openslide although I can use all tools like openslide-show-properties etc. successfully. CMake is also available. 
Could this mean that meson build script has some problem?

**Reply by jcupitt** [reply]
You can look at the meson logs to see exactly what went wrong. Perhaps it's failing to link to `libopenslide.so`? Check where you installed it, and verify that the linker can find it.

- if you installed to `/usr/local`, verify that `/usr/local/lib` is somewhere your linker is looking for libraries, and possibly `sudo ldconfig` to make it refresh the link cache
- if you installed to a prefix in your home directory, you'll need to add that lib area to your `LD_LIBRARY_PATH`

**Reply by tvngithub** [reply]
[meson-log.txt](https://github.com/libvips/libvips/files/14809113/meson-log.txt)
I've looked at the log file and it does indicate that the pkg-config command returns the version 4.0.0 but it's still flagged as 0 (not available) (see attached log file) . I've been searching around and could not figure out why the build is not taking openslide into account. Will continue to search.
BTW, another related question. 3DHistech has informed us that they will provide only DICOM format in the future. The underlying format is still MRXS but the file we will receive from the scanner will be DICOM. Will vips conversion still work to convert DICOM to Zoomify as of now with MRXS. Since the underlying format is still MRXS I could if necessary strip the DICOM information to retrieve MRXS but it is more convenient to use DICOM as input.
Thank you so much for your help.

**Reply by jcupitt** [reply]
The log says:

```
Called: `/usr/bin/pkg-config --cflags openslide` -> 1
stderr:
Package libdicom was not found in the pkg-config search path.
Perhaps you should add the directory containing `libdicom.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libdicom', required by 'openslide', not found
-----------
Pkg-config error with 'openslide': Could not generate cflags for openslide:
Package libdicom was not found in the pkg-config search path.
Perhaps you should add the directory containing `libdicom.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libdicom', required by 'openslide', not found
```

So your openslide install is missing libdicom (the DICOM load library). You need to install libdicom, then build openslide again.

https://github.com/ImagingDataCommons/libdicom

openslide has pretty good 3dhistech DICOM support, so you should be all set.

**Reply by jcupitt** [reply]
Thinking about this again, this is probably just the same library find issue. You're probably not setting up your lib area correctly.

What platform is this? I'll make you a demo dockerfile.

vipsdisp supports 3dhistech DICOM, so try some files with that to see if it works for you.

**Reply by tvngithub** [reply]
Thank your precise guidance I could compile VIPS with OpenSlide after building libdicom and added sqlite3 library. The current version 15.2 is super fast compared to the version that came with Ubuntu 22.04. I haven't benchmarked exact time it takes but is at least 2x faster than the current version on Ubuntu 22.04. This is great news for me and my application. I will try to use docker to install this build on all my servers running 22.04.

There is still one minor problem and that is the new VIPS dzsave stores jpg tiles with 4 channels while my application used jpg tiles with 3 channels in Zoomify format. Is there a setting or configuration that I can modify the jpeg output tiles? I could change the application (based on openCV) but if I can just produce the same output jpg tiles as before, that would be backward compatible.

Thanks again for your help. It's amazing that OpenSlide and VIPS continue to improve and more and more useful version after version. I appreciate the effort of dedicated developers on these projects.

**Reply by jcupitt** [reply]
I'm glad it's working.

You can use the `rgb` flag to make openslide read RGB pixels instead, eg.:

```
vips dzsave x.svs[rgb] y
```

To make `y.dzi` and `y_files`. It's quite a bit quicker too.

**Reply by tvngithub** [reply]
Thank you very much again.
With the option [rgb] the output is back like before and takes only 1/3 of the previous time (1m40s)
It is now even faster with the .dzi (1m20s) and _files (1m18s).
I am closing this this thread  and mark it as SOLVED.
